src/miner.cpp

```diff
@@ -270,13 +270,43 @@ ResVal<std::unique_ptr<CBlockTemplate>> BlockAssembler::CreateNewBlock(const CSc
     XVM xvm{};
     if (isEvmEnabledForBlock) {
         auto res = XResultValueLogged(evm_try_unsafe_construct_block_in_q(result, evmQueueId, pos::GetNextWorkRequired(pindexPrev, pblock->nTime, consensus), evmBeneficiary, blockTime, nHeight, static_cast<std::size_t>(reinterpret_cast<uintptr_t>(&mnview))));
-
-        auto r = XResultStatusLogged(evm_try_unsafe_remove_queue(result, evmQueueId));
-        if (!r) return Res::Err("Failed to remove queue");
-
         if (!res) return Res::Err("Failed to construct block");
         auto blockResult = *res;
+        if (!blockResult.failed_transaction.empty()) {
+            LogPrintf("%s: Construct block exceeded block size limit\n", __func__);
+            auto failedTxHash = std::string(blockResult.failed_transaction.data(), blockResult.failed_transaction.length());
+            auto res = XResultValueLogged(evm_try_unsafe_remove_txs_above_hash_in_q(result, evmQueueId, failedTxHash));
+            if (!res.ok) {
+                LogPrintf("%s: Unable to get txs to be removed from queue. Will result in a block hash mismatch.\n", __func__);
+            }
+            else {
+                auto failedTxHashes = *res;
+                std::set<uint256> evmTxsToUndo;
+                for (const auto &txStr : failedTxHashes) {
+                    auto txHash = std::string(txStr.data(), txStr.length());
+                    evmTxsToUndo.insert(uint256S(txHash));
+                }
 
+                // Get all EVM Txs
+                CTxMemPool::setEntries failedEVMTxs;
+                for (const auto& iter : inBlock) {
+                    auto tx = iter->GetTx();
+                    if (!evmTxsToUndo.count(tx.GetHash()))
+                        continue;
+                    std::vector<unsigned char> metadata;
+                    const auto txType = GuessCustomTxType(tx, metadata, false);
+                    if (txType == CustomTxType::EvmTx) {
```

Comment:
What about TransferDomain TXs?

Reply:
Only evmtxs have gas usage on the executor. This is guarded with the system_tx flag that is passed into the executor: https://github.com/DeFiCh/ain/pull/2526/files#diff-6f8e31143f7885918b267c7d971ca0193faeba1b27ef2229f2c8ee6ec4889336R190.

Reply:
We should not ever go into this loop: https://github.com/DeFiCh/ain/pull/2526/files#diff-6bf9d2e37d503f2f3381cd3a41c26e0b9670f26c2cd4571e98d2af9e3767c1a1R298-R300, and we should not be seeing theis log message.

Reply:
evm_try_unsafe_remove_txs_above_hash_in_q will remove TD TXs from the queue. If you could regenerate the block hash for the queue after removal for the EVM hash it would not match due to the block still having the TD TXs in.

All we are going to do here is set the miner up to create invalid blocks in a loop.

Reply:
This PR should not do that. To summarize, the pipeline is:
1. consensus builds the txqueue
2. construct block is called -> we do a final check here to ensure block doesnt exceed limit
3. if it somehow does, this safety layer catches it and fails that tx (and everything else that comes after)
4. kick out the failed evm txs from dvm block template

Transferdomain txs uses zero gas so that will never ever enter this safety net pipeline. EVM txs are never committed on the EVM block when they fail to add into the block due to the block size limit, and are returned as failed txs. Failed txs are subsequently removed from the block template on the miner, and will be included in the next block.

No invalid blocks should be minted since all txs are validated through consensus pipeline. EVM block hash mismatch will not happen either as long as we ensure proper handling of the txs to match correctly in both the DVM and EVM block.